rs/p2p/artifact_downloader/src/fetch_stripped_artifact/download.rs

```diff
@@ -0,0 +1,415 @@
+                        })
+                    {
+                        if IngressMessageId::from(&response.ingress_message) == ingress_message_id {
+                            return (response.ingress_message, peer);
+                        } else {
+                            warn!(
+                                log,
+                                "Peer {} responded with wrong artifact for advert", peer
+                            );
+                        }
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        sleep_until(next_request_at).await;
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use ic_logger::no_op_logger;
+    use ic_p2p_test_utils::mocks::{MockPeers, MockTransport, MockValidatedPoolReader};
+    use ic_test_utilities_consensus::{
+        fake::{Fake, FakeContentSigner},
+        make_genesis,
+    };
+    use ic_test_utilities_types::messages::SignedIngressBuilder;
+    use ic_types::batch::{BatchPayload, IngressPayload};
+    use ic_types::consensus::{dkg::Dealings, Block, BlockProposal, DataPayload, Payload, Rank};
+    use ic_types::Height;
+    use ic_types_test_utils::ids::{node_test_id, NODE_1};
+
+    fn mock_pools(
+        ingress_message: Option<SignedIngress>,
+        consensus_message: Option<ConsensusMessage>,
+    ) -> Pools {
+        let should_call_consensus_pool = ingress_message.is_none();
+
+        let mut ingress_pool = MockValidatedPoolReader::<SignedIngress>::default();
+        if let Some(ingress_message) = ingress_message {
+            ingress_pool
+                .expect_get()
+                .with(mockall::predicate::eq(IngressMessageId::from(
+                    &ingress_message,
+                )))
+                .once()
+                .return_const(ingress_message);
+        } else {
+            ingress_pool.expect_get().once().return_const(None);
+        }
+
+        let mut consensus_pool = MockValidatedPoolReader::<ConsensusMessage>::default();
+        if let Some(consensus_message) = consensus_message {
+            consensus_pool
+                .expect_get()
+                .with(mockall::predicate::eq(ConsensusMessageId::from(
+                    &consensus_message,
+                )))
+                .once()
+                .return_const(consensus_message.clone());
+        } else if should_call_consensus_pool {
+            consensus_pool.expect_get().once().return_const(None);
+        }
+
+        Pools {
+            consensus_pool: Arc::new(RwLock::new(consensus_pool)),
+            ingress_pool: Arc::new(RwLock::new(ingress_pool)),
+        }
+    }
+
+    #[tokio::test]
+    async fn rpc_get_from_ingress_pool_test() {
+        let ingress_message = SignedIngressBuilder::new().nonce(1).build();
+        let block = fake_block_proposal(vec![]);
+        let pools = mock_pools(Some(ingress_message.clone()), None);
+
+        let rpc_response = rpc_handler(
```

Comment:
I would have directly tested on the axum server here with `build_axum_router` and `Oneshot(request)`. This tests the more public interface and also catches issues with the router config.

Reply:
Good idea, done